modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+        Quat<_Tp> atanh() const;
+
+        /**
+         * @brief return true if this quaternion is a unit quaternion
+         * @param eps tolerance scope of normalization. The eps could be defined as
+         *
+         * \f[eps = |1 - dotValue|\f] where \f[dotValue = (this.w^2 + this.x^2 + this,y^2 + this.z^2)\f]
+         * And this function will consider it is normalized when the dotValude over a range \f$[1-eps, 1+eps]\f$.
+         */
+        bool isNormal(_Tp eps=CV_QUAT_EPS) const;
+
+        /**
+         * @brief to throw an error if this quaternion is not a unit quaternion
+         * @param eps tolerance scope of normalization
+         * @sa isNormal
+         */
+        void assertNormal(_Tp eps=CV_QUAT_EPS) const;
+
+        /**
+         * @brief transform a quaternion to a 3x3 rotate matrix.
+         * @param assumeUnit if true, this quaternon should be a unit quaternion and
+         * this function will save some computations. Otherwise, this function will normalize this
+         * quaternion at first then to do the transformation.
+         *
+         * @note Matrix A which is to be rotated has the form
+         * \f[\begin{bmatrix}
+         * x_0& x_1& x_2&...&x_n\\
+         * y_0& y_1& y_2&...&y_n\\
+         * z_0& z_1& z_2&...&z_n
+         * \end{bmatrix}\f]
+         * where the same subscript represents a point. The shape of A equals [3, n]
+         * The points matrix A can be rotated by toRotMat3x3() * A.
+         * The result has 3 rows and n columns too
+
+         * For example
+         * ```
+         * double angle = CV_PI;
+         * Vec3d axis{0,0,1};
+         * Quatd q_unit{angle, axis}; //quaternion coule also be get by interpolation by two or more quaternions.
+         *
+         * //assume there is two points (1,0,0) and (1,0,1) to be rotated
+         * Mat pointsA = (Mat_<double>(2, 3) << 1,0,0,1,0,1);
+         * //change the shape
+         * pointsA = pointsA.t()
+         * // rotate 180 degrees around the z axis
+         * Mat new_point = q_unit.toRotMat3x3() * pointsA;
+         * // print two points
+         * cout << new_point << endl;
+         * ```
+         */
+        cv::Mat toRotMat3x3(bool assumeUnit=false) const;
+
+        /**
+         * @brief transform a quaternion to a 4x4 rotate matrix.
+         * @param assumeUnit if true, this quaternon should be a unit quaternion and
+         * this function will save some computations. Otherwise, this function will normalize this
+         * quaternion at first then to do the transformation.
+         *
+         * The operations is similar as toRotMat3x3
+         * except that the points matrix should have the form
+         * \f[\begin{bmatrix}
+         * 0&0&0&...&0\\
+         * x_0& x_1& x_2&...&x_n\\
+         * y_0& y_1& y_2&...&y_n\\
+         * z_0& z_1& z_2&...&z_n
+         * \end{bmatrix}\f]
+         *
+         * @sa toRotMat3x3
+         */
+        cv::Mat toRotMat4x4(bool assumeUnit=false) const;
```

Comment:
The matrix layout is not compatible with the `Affine3`'s matrix layout (as well as with a lot of OpenCV functions from calib3d, see [this ](https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga41b1a8dd70eae371eba707d101729c36)for example).
It's the last row and the last column to contain zeros and ones, not the first ones as in your code.

Reply:
OK